### PR TITLE
fix(service-food-crawler): ulf support, includes crawler refactor

### LINKED ISSUE #44: Crawling Ulf's Café not working
**Describe the bug**

image

**Environment:**

 - OS:  
 - hpi-cloud version:

### PR DESCRIPTION
**Pull request by ShirleyNekoDev**
**Fixes: #44**

**Checklist**

- [x] Commit messages follow conventionalcommits.org
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Changes**

- adds utils in common
- refactor parser customization
  - label parsing
  - category parsing
- support for Ulfs cafe

### CODE REVIEW

File: common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt [hunk 1]
```diff
@@ -20,10 +20,16 @@ fun <E : Any> MutableList<out E>.removeFirst(filter: (E) -> Boolean): E? {
         elem
     } else null
 }
+
+fun <T: Any> T.asSingletonList(): List<T> = listOf(this)
```
Comment (JonasWanke) [inline common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt]: This isn't really required but makes our code more confusing for new developers

File: common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt [hunk 2]
```diff
@@ -20,10 +20,16 @@ fun <E : Any> MutableList<out E>.removeFirst(filter: (E) -> Boolean): E? {
...
 
 // region Map
 fun <K, V> mapOfNotNull(vararg pairs: Pair<K, V>?): Map<K, V> {
     return pairs.filterNotNull().toMap()
 }
-// endregion
\ No newline at end of file
+
+fun <K, V> Map<K, V?>.getOrElse(key: K, alternativeValue: () -> V?) =
+    if(this.containsKey(key)) this[key]
```
Comment (JonasWanke) [inline common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt]: Missing a space between `if` and opening bracket

File: service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package de.hpi.cloud.food.crawler.canteens
+
+import de.hpi.cloud.food.crawler.CanteenData
+import de.hpi.cloud.food.crawler.MensaMeal
+
+object Ulf : CanteenData("ulfsCafe", 112) {
+    override fun mapReduce(meals: List<MensaMeal>): List<MensaMeal> = meals
+        // ask Kotlin why I have to specify the type for R here
```
Comment (JonasWanke) [inline service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: Do you mean the return type `: List<MensaMeal>`? Not specifying it works for me

File: service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        .flatMap {
+            val alternativeOffer = it.openMensaMeal.notes
+                .joinToString(separator = " ")
+                .trim()
+            if (alternativeOffer.startsWith("oder ", ignoreCase = true)) {
+                listOf(it, it.copy(title = alternativeOffer.substring(5)))
+            } else
+                listOf(it)
```
Comment (JonasWanke) [inline service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: One branch with curly braces and one without?

**Review by JonasWanke — CHANGES_REQUESTED** [review]

File: common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt (hunk 1 above)
Comment (ShirleyNekoDev) [reply common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt]: I don't think this is a bad pattern. I'd say it's more readable as listOf(object).
Would you prefer .let{listOf(it)} ?

File: service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt (hunk 3 above)
Comment (ShirleyNekoDev) [reply service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: What do you mean? Shouldn't be a problem, but I can remove the return type specifier if you want.

File: service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt (hunk 4 above)
Comment (ShirleyNekoDev) [reply service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: oops

File: common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt (hunk 1 above)
Comment (JonasWanke) [reply common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt]: Yes, that would be easier to understand

File: service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt (hunk 3 above)
Comment (JonasWanke) [reply service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: I'm not sure what your comment in code refers to

Comment (ShirleyNekoDev) [reply service-food-crawler/src/main/kotlin/de/hpi/cloud/food/crawler/canteens/Ulf.kt]: Nvm. This is old and can be removed.

File: common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt (hunk 1 above)
Comment (ShirleyNekoDev) [reply common/src/main/kotlin/de/hpi/cloud/common/utils/Collections.kt]: Sure, no problem

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**